rfd/0143-external-k8s-joining.md

```diff
@@ -0,0 +1,280 @@
+  }
+  
+  // Clusters is a list of Cluster that are accepted sources of JWTs to use
+  // against this token. A Service Account JWT signed by a cluster not listed
+  // here will be rejected.
+  repeated Cluster Clusters = 1;
+  // Allow is a list of Rules, nodes using this token must match one
+  // allow rule to use this token.
+  repeated Rule Allow = 2;
+}
+```
+
+In order to obtain the `jwks` value for their Kubernetes Cluster, operators
+will need to run the following steps from a machine already authenticated
+with their cluster:
+
+- `kubectl proxy -p 8080`
+- `curl http://localhost:8080/openid/v1/jwks`
+
+In the case that an operator rotates the certificate authority used by their
+Kubernetes Cluster to sign Service Account JWTS, the `jwks` field will also
+need to be updated.
+
+#### Configuring a pod for `kubernetes-jwks` joining
+
+The Pod will require a service account with a role granting it the ability to
+call the TokenRequest endpoint for itself. This allows the creation of a token
+with an audience of our choosing in order to include a nonce during the join
+to prevent token reuse.
+
+E.g for a service account named: "argocd":
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
+metadata:
+  name: argocd-token-request
+rules:
+- apiGroups: [""]
+  resources:
+  - "serviceaccounts/token"
+  verbs:
+  - "create"
+  # Only grant the ability to create tokens for itself. Missing this field leads
+  # to a dangerously powerful role.
+  resourceNames:
+    - argocd
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: RoleBinding
+metadata:
+  name: argocd-token-review
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: argocd-token-review
+subjects:
+- kind: ServiceAccount
+  name: argocd
+  namespace: default
+```
+
+### Implementation
+
+<-- TODO: Specs for RegisterUsingKubernetesJWKS RPC -->
+
+The `kubernetes-jwks` flow will leverage a challenge and response flow, similar
+to that implemented for IAM joining. The flow will run as follows:
+
+1. The client wishing to join calls the `RegisterUsingKubernetesJWKS` RPC. It
+  submits the name of the token it wants to use.
+2. The Auth Server sends a response with the audience that should be set in the 
+  token. This audience is a concatenation of the cluster name, and 24 bytes of
+  cryptographically random data encoded in base64 URL encoding, e.g:
+  `example.teleport.sh/YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4`
+3. The client uses its Kubernetes Service Account to call TokenRequest. It
+  requests a JWT for its own service account, with the audience specified by
+  the Auth Server and with a short time-to-live.
+4. The client sends a further request up the stream including this token.
+5. The server validates:
+  a. The JWT is signed by the keys present in the `jwks` field of the token 
+    resource.
+  b. The JWT is not expired, and at the time of issue, the JWTs TTL was short
+   (e.g within a similar TTL to the 30s TTL we set)
```

Comment:
Maybe 30s is a bit short. I'm afraid of the join method exploding as soon as a master node has a clock skew.

Reply:
I still prefer to use short time periods. Drifting the clocks will cause issues with validating certificates either way so it's better to abort anw